nautobot/core/views/mixins.py

```diff
@@ -0,0 +1,804 @@
+                form = self._handle_object_does_not_exist(form, self.logger)
+        elif self.action == "bulk_create":
+            try:
+                self.obj_table = self._process_bulk_create_form(form)
+            except ValidationError:
+                self.has_error = True
+                pass
+            except ObjectDoesNotExist:
+                form = self._handle_object_does_not_exist(form, self.logger)
+
+        if not self.has_error:
+            self.logger.debug("Form validation was successful")
+            if self.action == "bulk_create":
+                return Response(
+                    {
+                        "table": self.obj_table,
+                        "template": "import_success.html",
+                    }
+                )
+            return super().form_valid(form)
+        else:
+            data = {}
+            if self.action in ["bulk_update", "bulk_destroy"]:
+                pk_list = self.request.POST.getlist("pk")
+                table = self.table_class(self.queryset.filter(pk__in=pk_list), orderable=False)
+                if not table.rows:
+                    messages.warning(
+                        request,
+                        f"No {self.queryset.model._meta.verbose_name_plural} were selected for {self.action}.",
+                    )
+                    return redirect(self.get_return_url(request))
+
+                data.update({"table": table})
+            data.update({"form": form})
+            return Response(data)
+
+    def form_invalid(self, form):
+        """
+        Handle invalid forms.
+        """
+        data = {}
+        request = self.request
+        if self.action in ["bulk_update", "bulk_destroy"]:
+            pk_list = self.request.POST.getlist("pk")
+            table = self.table_class(self.queryset.filter(pk__in=pk_list), orderable=False)
+            if not table.rows:
+                messages.warning(
+                    request,
+                    f"No {self.queryset.model._meta.verbose_name_plural} were selected for {self.action}.",
+                )
+                return redirect(self.get_return_url(request))
+
+            data = {
+                "table": table,
+            }
+        data.update({"form": form})
+        return Response(data)
+
+    def get_object(self):
+        """
+        Returns the object the view is displaying.
+        You may want to override this if you need to provide non-standard
+        queryset lookups.  Eg if objects are referenced using multiple
+        keyword arguments in the url conf.
+        """
+        queryset = self.queryset
+        # Perform the lookup filtering.
+        lookup_url_kwarg = self.lookup_url_kwarg or self.lookup_field
+        if lookup_url_kwarg not in self.kwargs:
+            return self.queryset.model()
+        filter_kwargs = {self.lookup_field: self.kwargs[lookup_url_kwarg]}
+        obj = get_object_or_404(queryset, **filter_kwargs)
+
+        return obj
```

Comment:
I would like to see if we can remove this and just the base [`get_object()` that is inherited from `GenericViewSet`](https://github.com/encode/django-rest-framework/blob/master/rest_framework/generics.py#L75-L101).

Reply:
the base `get_object()` calls `check_object_permissions()` which I had to override and make it exactly the same as `check_permissions()` in this file. This will result in a double database hit per request, see Glenn's [comment](https://github.com/nautobot/nautobot/pull/1902#discussion_r942508858). So I decided to override `get_object()` and to remove the line that calls `check_object_permissions()`. Since `check_permissions()` has already covered object-level permissions and I dont have to redefine `check_object_permissions()`.

Reply:
I tried that but filter_queryset() in the default get_object() throws an error when we tried to add ?tab=main in any of the objectdetailview url querystrings. So I think it would still be much easier to just override the get_object() method itself to fit our needs. We may run into more problems down the road when we try to use multiple keywords to get an object (see docstring)
Screen Shot 2022-08-12 at 3 21 57 PM